### ISSUE TITLE
Segmentation fault when making tvalues.m2 on multiple architectures

### ISSUE THREAD

**Issue by d-torrance** [Macaulay2/M2#1554]
We're getting the following build error when building Macaulay2 on several Debian architectures:

```shell
cp .gdbinit.buildd .gdbinit
echo '# -*- sh -*-' >.gdb-files
/bin/echo 'echo -- loading .gdb-files\n' >>.gdb-files
echo 'set environment LD_LIBRARY_PATH /<<PKGBUILDDIR>>/M2/usr-host/lib:' >>.gdb-files
echo 'file /<<PKGBUILDDIR>>/M2/usr-dist/aarch64-Linux-Debian-unknown/bin/M2-binary' >>.gdb-files
/bin/echo 'echo -- loaded .gdb-files\n' >>.gdb-files
/<<PKGBUILDDIR>>/M2/usr-dist/aarch64-Linux-Debian-unknown/bin/M2 -q --silent --stop -e errorDepth=0 --no-preload --no-tvalues tvalues-make.m2 -e 'make "./../d/"; exit 0'
make[3]: *** [Makefile:43: tvalues.m2] Segmentation fault
make[3]: Leaving directory '/<<PKGBUILDDIR>>/M2/Macaulay2/m2'
make[2]: *** [Makefile:14: all-in-m2] Error 2
make[2]: Leaving directory '/<<PKGBUILDDIR>>/M2/Macaulay2'
make[1]: *** [GNUmakefile:229: all-in-Macaulay2] Error 2
make[1]: Leaving directory '/<<PKGBUILDDIR>>/M2'
```

Here are the full build logs for each affected architecture:

* arm64: https://buildd.debian.org/status/fetch.php?pkg=macaulay2&arch=arm64&ver=1.16%2Bgit55.94c4b7d%2Bds-1&stamp=1603488826&raw=0
* armel: https://buildd.debian.org/status/fetch.php?pkg=macaulay2&arch=armel&ver=1.16%2Bgit55.94c4b7d%2Bds-1&stamp=1603489172&raw=0
* armhf: https://buildd.debian.org/status/fetch.php?pkg=macaulay2&arch=armhf&ver=1.16%2Bgit55.94c4b7d%2Bds-1&stamp=1603489340&raw=0
* ppc64el: https://buildd.debian.org/status/fetch.php?pkg=macaulay2&arch=ppc64el&ver=1.16%2Bgit55.94c4b7d%2Bds-1&stamp=1603488033&raw=0
* s390x: https://buildd.debian.org/status/fetch.php?pkg=macaulay2&arch=s390x&ver=1.16%2Bgit55.94c4b7d%2Bds-1&stamp=1603487881&raw=0
* powerpc: https://buildd.debian.org/status/fetch.php?pkg=macaulay2&arch=powerpc&ver=1.16%2Bgit55.94c4b7d%2Bds-1&stamp=1603488777&raw=0
* ppc64: https://buildd.debian.org/status/fetch.php?pkg=macaulay2&arch=ppc64&ver=1.16%2Bgit55.94c4b7d%2Bds-1&stamp=1603488550&raw=0
* riscv64: https://buildd.debian.org/status/fetch.php?pkg=macaulay2&arch=riscv64&ver=1.16%2Bgit55.94c4b7d%2Bds-1&stamp=1603491909&raw=0

Note that several of these are official Debian architectures, i.e., these build failures will block Macaulay2 from migrating from unstable to testing (and therefore, eventually to stable).

I recall seeing similar build failures a few months ago in the Github workflow builds.  How were they fixed?  Is there a commit I can cherry-pick for the Debian package?  (It's based on the `release-1.16` branch.)

**Comment by mahrud** [comment]
It might be related to #1370 which is still open, though it's hard to tell without a stacktrace, and that was only impacting CMake builds on ubuntu. That issue also happened when extracting the typical values.

**Comment by d-torrance** [comment]
Not sure if this is the same underlying bug, but I'm starting to get a segfault at this point on local builds, too.  It appears to be some issue with blas and threads.  Here's the output from gdb:

```gdb
(gdb) r
Starting program: /home/profzoom/src/macaulay2/M2/M2/BUILD/doug/usr-dist/x86_64-Linux-Ubuntu-20.10/bin/M2-binary 
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
[New Thread 0x7ffff1929640 (LWP 515909)]
[New Thread 0x7ffff1128640 (LWP 515910)]
[New Thread 0x7fffe0927640 (LWP 515911)]
[New Thread 0x7fffe0126640 (LWP 515912)]
[New Thread 0x7fffcf925640 (LWP 515913)]
[New Thread 0x7fffc7124640 (LWP 515914)]
[New Thread 0x7fffbe923640 (LWP 515915)]
[New Thread 0x7fffb6122640 (LWP 515916)]
[New Thread 0x7fffad921640 (LWP 515917)]
[New Thread 0x7fffa5120640 (LWP 515918)]
[New Thread 0x7fff9c91f640 (LWP 515919)]
[New Thread 0x7fff9c0ee640 (LWP 515920)]

Thread 12 "M2-binary" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7fff9c91f640 (LWP 515919)]
0x0000000000000000 in ?? ()
(gdb) bt
#0  0x0000000000000000 in ?? ()
#1  0x00007ffff40902a9 in blas_memory_alloc (procpos=procpos@entry=2)
    at memory.c:2787
#2  0x00007ffff4090aa8 in blas_thread_server (arg=<optimized out>)
    at blas_server.c:327
#3  0x00007ffff655a590 in start_thread (arg=0x7fff9c91f640)
    at pthread_create.c:463
#4  0x00007ffff601d223 in clone ()
    at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
(gdb) 
```

And then the really strange thing -- if I run my already-installed copy of M2 in the same terminal as the build, then I get the same segfault.  Once I switch to another terminal, it's fine.

**Comment by DanGrayson** [comment]
Which blas library is multi-threaded?

**Comment by d-torrance** [comment]
> Which blas library is multi-threaded?

OpenBLAS is.  For some reason, my PPA builds are linking against it, too:

```
dtorrance@nimmie:~$ ldd /usr/bin/M2-binary | grep blas
	libblas.so.3 => /usr/lib/x86_64-linux-gnu/libblas.so.3 (0x00007f40f5135000)
	libopenblas.so.0 => /usr/lib/x86_64-linux-gnu/libopenblas.so.0 (0x00007f40f15ee000)
```

I'm thinking this may be a separate issue than the non-amd64 build failures for the Debian package, as OpenBLAS isn't showing up anywhere in the build logs.

**Comment by DanGrayson** [comment]
Is there any possibility of getting logins on machines with all of those architectures?  Then we could build debug versions and use gdb to see what's going on.

**Comment by d-torrance** [comment]
> Is there any possibility of getting logins on machines with all of those architectures? Then we could build debug versions and use gdb to see what's going on.

It is is possible -- Debian has a bunch of "porterboxes" that developers can use for this purpose.  I'll look into it -- I need to get sponsored by a Debian Developer to get access.

**Comment by DanGrayson** [comment]
Add my name, too, in case I need to take a look at some of them.

**Comment by d-torrance** [comment]
I'm still waiting on an update for porterbox access, but in the meantime, I set up an arm64 chroot using qemu to investigate a bit more.

Unfortunately, gdb is very unhelpful:

```gdb
(sid-arm64)root@nimmie:~/M2/M2/BUILD/doug# usr-dist/aarch64-Linux-Debian-unstable/bin/M2
qemu: uncaught target signal 11 (Segmentation fault) - core dumped
Segmentation fault
(sid-arm64)root@nimmie:~/M2/M2/BUILD/doug# gdb -c qemu_M2-binary_20201124-164654_1973591.core 
GNU gdb (Debian 10.1-1+b1) 10.1
Copyright (C) 2020 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "aarch64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word".
[New LWP 1973591]
[New LWP 1973601]
[New LWP 1973602]
[New LWP 1973603]
Core was generated by `e�U  '.
#0  0x0000005500108584 in ?? ()
[Current thread is 1 (LWP 1973591)]
(gdb) bt
#0  0x0000005500108584 in ?? ()
#1  0x00000055005060d8 in ?? ()
Backtrace stopped: previous frame inner to this frame (corrupt stack?)
```

**Comment by DanGrayson** [comment]
When I want to use gdb for debugging, I ensure that the executable retains its symbol table -- normally it is stripped, and I compile with the -g3 option.  Just add --enable-debug to the configure command line.

Or did you do that already?

**Comment by d-torrance** [comment]
> When I want to use gdb for debugging, I ensure that the executable retains its symbol table -- normally it is stripped, and I compile with the -g3 option. Just add --enable-debug to the configure command line.
> 
> Or did you do that already?

Yes, I did.  :confused:

**Comment by DanGrayson** [comment]
Very strange.

Is qemu really fast enough?

**Comment by d-torrance** [comment]
> Very strange.
> 
> Is qemu really fast enough?

I'm not sure -- this is my first time really playing with it.   One big disadvantage is that I can't use gdb directly on the binary itself because qemu doesn't support ptrace:

```gdb
(sid-arm64)root@peg-amy:~/macaulay2-1.16+git55.94c4b7d+ds/M2# gdb usr-dist/aarch64-Linux-Debian-unknown/bin/M2-binary
GNU gdb (Debian 10.1-1+b1) 10.1
Copyright (C) 2020 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "aarch64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from usr-dist/aarch64-Linux-Debian-unknown/bin/M2-binary...
(gdb) r
Starting program: /root/macaulay2-1.16+git55.94c4b7d+ds/M2/usr-dist/aarch64-Linux-Debian-unknown/bin/M2-binary 
warning: Could not trace the inferior process.
warning: ptrace: Function not implemented
During startup program exited with code 127.
```
I can only use gdb on a core dump as above.

I just ordered a Raspberry Pi for a class I'm teaching next semester.  Once it arrives, I'm going to try building Macaulay2 on it.  Hopefully that will be helpful!

**Comment by DanGrayson** [comment]
Once before we built it on a raspberry pi.  It made an impressive demo at a workshop.

**Comment by d-torrance** [comment]
The Raspberry Pi arrived!  These are the results from my first attempt at building and running gdb:

```gdb
(sid-armhf)root@raspberrypi:~/macaulay2-1.16+git55.94c4b7d+ds# gdb M2/usr-dist/arm-Linux-Debian-unknown/bin/M2-binary
GNU gdb (Debian 10.1-1+b1) 10.1
Copyright (C) 2020 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "arm-linux-gnueabihf".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from M2/usr-dist/arm-Linux-Debian-unknown/bin/M2-binary...
(gdb) run
Starting program: /root/macaulay2-1.16+git55.94c4b7d+ds/M2/usr-dist/arm-Linux-Debian-unknown/bin/M2-binary 
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/arm-linux-gnueabihf/libthread_db.so.1".
[New Thread 0xb3ecf7b0 (LWP 17763)]
[New Thread 0xb36ce7b0 (LWP 17764)]
[New Thread 0xb2ecd7b0 (LWP 17765)]

Thread 1 "M2-binary" received signal SIGSEGV, Segmentation fault.
0x004a0756 in lex_makeUniqueWord (s=0xb3ee98b8, p=0xb3ee6390) at lex.d:20
20	     hashList := hashTable.hashCode;
(gdb) thread apply all bt

Thread 4 (Thread 0xb2ecd7b0 (LWP 17765) "M2-binary"):
#0  0xb6020664 in __libc_do_syscall () from /lib/arm-linux-gnueabihf/libpthread.so.0
#1  0xb601b114 in pthread_cond_wait@@GLIBC_2.4 () from /lib/arm-linux-gnueabihf/libpthread.so.0
#2  0xb6069fb4 in GC_wait_marker () from /usr/lib/arm-linux-gnueabihf/libgc.so.1
#3  0xb606270e in GC_help_marker () from /usr/lib/arm-linux-gnueabihf/libgc.so.1
#4  0xb6069f5e in GC_mark_thread () from /usr/lib/arm-linux-gnueabihf/libgc.so.1
#5  0xb601598e in start_thread () from /lib/arm-linux-gnueabihf/libpthread.so.0
#6  0xb5dfaaec in ?? () from /lib/arm-linux-gnueabihf/libc.so.6
Backtrace stopped: previous frame identical to this frame (corrupt stack?)

Thread 3 (Thread 0xb36ce7b0 (LWP 17764) "M2-binary"):
#0  0xb6020664 in __libc_do_syscall () from /lib/arm-linux-gnueabihf/libpthread.so.0
#1  0xb601b114 in pthread_cond_wait@@GLIBC_2.4 () from /lib/arm-linux-gnueabihf/libpthread.so.0
#2  0xb6069fb4 in GC_wait_marker () from /usr/lib/arm-linux-gnueabihf/libgc.so.1
--Type <RET> for more, q to quit, c to continue without paging--
#3  0xb606270e in GC_help_marker () from /usr/lib/arm-linux-gnueabihf/libgc.so.1
#4  0xb6069f5e in GC_mark_thread () from /usr/lib/arm-linux-gnueabihf/libgc.so.1
#5  0xb601598e in start_thread () from /lib/arm-linux-gnueabihf/libpthread.so.0
#6  0xb5dfaaec in ?? () from /lib/arm-linux-gnueabihf/libc.so.6
Backtrace stopped: previous frame identical to this frame (corrupt stack?)

Thread 2 (Thread 0xb3ecf7b0 (LWP 17763) "M2-binary"):
#0  0xb6020664 in __libc_do_syscall () from /lib/arm-linux-gnueabihf/libpthread.so.0
#1  0xb601b114 in pthread_cond_wait@@GLIBC_2.4 () from /lib/arm-linux-gnueabihf/libpthread.so.0
#2  0xb6069fb4 in GC_wait_marker () from /usr/lib/arm-linux-gnueabihf/libgc.so.1
#3  0xb606270e in GC_help_marker () from /usr/lib/arm-linux-gnueabihf/libgc.so.1
#4  0xb6069f5e in GC_mark_thread () from /usr/lib/arm-linux-gnueabihf/libgc.so.1
#5  0xb601598e in start_thread () from /lib/arm-linux-gnueabihf/libpthread.so.0
#6  0xb5dfaaec in ?? () from /lib/arm-linux-gnueabihf/libc.so.6
Backtrace stopped: previous frame identical to this frame (corrupt stack?)

--Type <RET> for more, q to quit, c to continue without paging--
Thread 1 (Thread 0xb3f11010 (LWP 17760) "M2-binary"):
#0  0x004a0756 in lex_makeUniqueWord (s=0xb3ee98b8, p=0xb3ee6390) at lex.d:20
#1  0x0044cbf2 in nleftword (s=<optimized out>) at binding.d:162
#2  0x0044eef4 in binding__prepare () at binding.d:204
#3  0x006a7702 in __libc_csu_init ()
#4  0xb5d749dc in __libc_start_main () from /lib/arm-linux-gnueabihf/libc.so.6
#5  0x004833e8 in _start ()
Backtrace stopped: previous frame identical to this frame (corrupt stack?)
(gdb) 
```

**Comment by DanGrayson** [comment]
An int is 32 bits, right?

**Comment by d-torrance** [comment]
> An int is 32 bits, right?

Yes.  `long` is the one that's not consistent between architectures.  Here's a nice rundown for the various Debian architectures:
https://wiki.debian.org/ArchitectureSpecificsMemo

**Comment by d-torrance** [comment]
> Here's a nice rundown for the various Debian architectures:
> https://wiki.debian.org/ArchitectureSpecificsMemo

I just noticed that all of the architectures that are segfaulting have unsigned `char`, while the architectures where Macaulay2 builds (or at least, the build doesn't fail until the example-generating/testing phase...) have signed `char`.  Perhaps that's the issue?

**Comment by d-torrance** [comment]
I tried compiling on the Raspberry Pi with `--enable-debug` passed to `configure`, and that's shedding some more light on the situation:

```m2
(macaulay2-armhf)root@raspberrypi:~/M2/M2/BUILD/doug# usr-dist/armv7l-Linux-Debian-unknown/bin/M2-binary 
../../../../Macaulay2/d/lex.d:20:27: error: array index -3003 out of bounds 0 .. 7312
-* stack trace, pid: 7307
 0# stack_trace(std::ostream&, bool) at ../../../../Macaulay2/d/main.cpp:124
-- end stack trace *-
```
Here are the lines where the problem is occurring:
https://github.com/Macaulay2/M2/blob/8ecbb08b154be4e1a39e83faf8576112311e322d/M2/Macaulay2/d/lex.d#L17-L20

So `hash` must be spitting out a negative value.  Here's the code for the hashing function:
https://github.com/Macaulay2/M2/blob/8ecbb08b154be4e1a39e83faf8576112311e322d/M2/Macaulay2/d/strings.d#L38-L42

Perhaps something funky is going on with `int(c)` casting an unsigned `char` to a signed `int`?

**Comment by d-torrance** [comment]
I tried adding a line to `strings.d` printing the values of `c` and `h` during each iteration of the `foreach` loop, and we're indeed getting a negative hash:

```c
c = -, h = 45
c = *, h = 1437
c = e, h = 44648
c = n, h = 1384198
c = d, h = 42910238
c =  , h = 1330217410
c = o, h = -1712933139
c = f, h = -1561319655
c =  , h = -1156269017
c = f, h = -1484601057
c = i, h = 1222007594
c = l, h = -772470142
c = e, h = 1823229475
c = *, h = 685538919
c = -, h = -223129946
```
`-223129946 % 7313 = -3003`, so that's what's going on.

**Comment by d-torrance** [comment]
Actually, that's not the problem.  My amd64 system also prints -223129946 as the final value of `h`.  But that final bitwise AND with `0x7fffffff` isn't doing anything on the Raspberry Pi:

amd64
------
```c
c = -, h = 45
c = *, h = 1437
c = e, h = 44648
c = n, h = 1384198
c = d, h = 42910238
c =  , h = 1330217410
c = o, h = -1712933139
c = f, h = -1561319655
c =  , h = -1156269017
c = f, h = -1484601057
c = i, h = 1222007594
c = l, h = -772470142
c = e, h = 1823229475
c = *, h = 685538919
c = -, h = -223129946
final hash = 1924353702
```

armhf
-----
```c
c = *, h = 1437
c = e, h = 44648
c = n, h = 1384198
c = d, h = 42910238
c =  , h = 1330217410
c = o, h = -1712933139
c = f, h = -1561319655
c =  , h = -1156269017
c = f, h = -1484601057
c = i, h = 1222007594
c = l, h = -772470142
c = e, h = 1823229475
c = *, h = 685538919
c = -, h = -223129946
final hash = -223129946
```

**Comment by d-torrance** [comment]
It looks like ARM uses two's complement for signed integers [1], so ANDing with `0x7fffffff` isn't the correct way to make the hash positive in the end.

I suppose we could detect one's v. two's complement during configure, add a function somewhere that does the correct thing depending on the implementation, and call that everywhere we're currently masking out the signed bit.

**Comment by DanGrayson** [comment]
Two's complement is what all our architectures use, anyway, and for both one's and two's complement, clearing the top bit makes the number positive.

**Comment by d-torrance** [comment]
> Two's complement is what all our architectures use, anyway, and for both one's and two's complement, clearing the top bit makes the number positive.

D'oh!  I was so excited that I may have figured it out that I didn't notice that x86 was the first one mentioned in the list of two's complement architectures on that Wikipedia page.  :)

But certainly there's something different -- ANDing with `0x7fffffff` isn't doing anything on ARM.

**Comment by d-torrance** [comment]
It does work as expected in a simple C program:

```c
#include <stdio.h>

int main(void)
{
        printf("%d\n", -223129946 & 0x7fffffff);
        return 0;
}
```

On both architectures, it's printing `1924353702`.

**Comment by DanGrayson** [comment]
> But certainly there's something different -- ANDing with `0x7fffffff` isn't doing anything on ARM.

That's why I asked whether an "int" was 32 bits.  Can it really be doing nothing?  Maybe single step one machine instruction at a time and see what's in the registers.  Or look at the assembly output from the compiler.

**Comment by d-torrance** [comment]
It must have something to do with how the variable `h` is being represented bitwise.  I added the following lines to `strings.d` after the above for loop:

```m2
 Ccode(void, "printf(\"using literal value of h: %d\\n\", -223129946 & 0x7fffffff);");
 Ccode(void, "printf(\"using variable h: %d\\n\", h & 0x7fffffff);");
```
 
and got:
```m2
using literal value of h: 1924353702
using variable h: -223129946
```

**Comment by DanGrayson** [comment]
Use "-S" to get the assembler listing from the file strings-tmp.c and let's see the code for the function strings_hash.  Or use gdb to disassemble it, as well as any constants it refers to.

**Comment by d-torrance** [comment]
> Use "-S" to get the assembler listing from the file strings-tmp.c and let's see the code for the function strings_hash.

```assembly
strings_hash:
	@ args = 0, pretend = 0, frame = 24
	@ frame_needed = 1, uses_anonymous_args = 0
	@ link register save eliminated.
	push	{r7}
	sub	sp, sp, #28
	add	r7, sp, #0
	str	r0, [r7, #4]
	movs	r3, #0
	str	r3, [r7, #20]
	movs	r3, #1
	str	r3, [r7, #12]
	ldr	r3, [r7, #4]
	str	r3, [r7, #8]
	ldr	r3, [r7, #8]
	ldr	r3, [r3]
	cmp	r3, #0
	beq	.L169
	ldr	r3, [r7, #12]
	cmp	r3, #0
	ble	.L163
	movs	r3, #0
	str	r3, [r7, #16]
	b	.L167
.L163:
	ldr	r3, [r7, #8]
	ldr	r3, [r3]
	subs	r3, r3, #1
	str	r3, [r7, #16]
.L164:
.L167:
	ldr	r2, [r7, #20]
	mov	r3, r2
	lsls	r3, r3, #5
	subs	r3, r3, r2
	ldr	r1, [r7, #8]
	ldr	r2, [r7, #16]
	add	r2, r2, r1
	adds	r2, r2, #4
	ldrb	r2, [r2]	@ zero_extendqisi2
	add	r3, r3, r2
	str	r3, [r7, #20]
	ldr	r2, [r7, #16]
	ldr	r3, [r7, #12]
	add	r3, r3, r2
	str	r3, [r7, #16]
	ldr	r3, [r7, #12]
	cmp	r3, #0
	ble	.L165
	ldr	r3, [r7, #8]
	ldr	r3, [r3]
	ldr	r2, [r7, #16]
	cmp	r2, r3
	blt	.L167
.L165:
	ldr	r3, [r7, #12]
	cmp	r3, #0
	bge	.L170
	ldr	r3, [r7, #16]
	cmp	r3, #0
	blt	.L170
	b	.L167
.L169:
	nop
	b	.L162
.L170:
	nop
.L162:
	ldr	r3, [r7, #20]
	bic	r3, r3, #-2147483648
	mov	r0, r3
	adds	r7, r7, #28
	mov	sp, r7
	@ sp needed
	ldr	r7, [sp], #4
	bx	lr
	.size	strings_hash, .-strings_hash
	.align	1
	.global	strings_index
	.syntax unified
	.thumb
	.thumb_func
	.fpu vfpv3-d16
	.type	strings_index, %function
```

**Comment by d-torrance** [comment]
I'm not seeing any difference in bits when I print the registers that hold the value of `h` in binary using gdb:

# armhf
```gdb
(gdb) p /t $r4
$1 = 11110010101100110100111010100110
```

# amd64
```gdb
(gdb) p /t $rbx
$2 = 11110010101100110100111010100110
```

**Comment by DanGrayson** [comment]
```
	bic	r3, r3, #-2147483648
	mov	r0, r3
```

The "bic" instruction above clears the sign bit, and the "mov" instruction puts the number in r0 as the return value from the function.  You should be able to watch it happen in gdb, by single stepping.  And then see the function that called it using a positive hash code.

It would be interesting to see the 32 bits of the bic instruction, to see if it got assembled correctly.

**Comment by d-torrance** [comment]
Oops, the above was just the output of `gcc -S`, with no other build flags.  Running `gdb` on `M2-binary` gives us something a little different, I guess due to optimization:

```asm
│   0x497f00 <strings_hash>                 push    {r4}                       │
│   0x497f02 <strings_hash+2>               ldr     r4, [r0, #0]               │
│   0x497f04 <strings_hash+4>               cbz     r4, 0x497f28 <strings_hash │
│   0x497f06 <strings_hash+6>               mov     r12, r0                    │
│   0x497f08 <strings_hash+8>               adds    r2, r0, #3                 │
│   0x497f0a <strings_hash+10>              movs    r0, #0                     │
│   0x497f0c <strings_hash+12>              mov     r3, r2                     │
│   0x497f0e <strings_hash+14>              ldrb.w  r1, [r2, #1]!              │
│   0x497f12 <strings_hash+18>              subs    r3, #2                     │
│   0x497f14 <strings_hash+20>      rsb     r0, r0, r0, lsl #5                 │
│   0x497f18 <strings_hash+24>      sub.w   r3, r3, r12                        │
│   0x497f1c <strings_hash+28>      add     r0, r1                             │
│   0x497f1e <strings_hash+30>      cmp     r4, r3                             │
│   0x497f20 <strings_hash+32>      bgt.n   0x497f0c <strings_hash+12>         │
│   0x497f22 <strings_hash+34>      ldr.w   r4, [sp], #4                       │
│   0x497f26 <strings_hash+38>      bx      lr                                 │
│   0x497f28 <strings_hash+40>      mov     r0, r4                             │
│   0x497f2a <strings_hash+42>      ldr.w   r4, [sp], #4                       │
│   0x497f2e <strings_hash+46>      bx      lr                                 │
└

**Comment by DanGrayson** [comment]
I think the clearing of the sign bit got optimized away.  Get the command line from "make" for compiling that file, add -O0 to it, recompile, relink, and see if the bug is gone.

Maybe the optimizer assumes incorrectly that our arithmetic operations never result in an overflow.

Yes!!  It does!!!

"an annoying aspect of the C11 standard, namely that signed-integer overflow is undefined"

See https://lwn.net/Articles/511259/

**Comment by DanGrayson** [comment]
If that's correct, we just have to switch to unsigned ints, where a carry into the top bit is not regarded as an overflow:
```diff
diff --git a/M2/Macaulay2/d/strings.d b/M2/Macaulay2/d/strings.d
index 166bb4b32..8261aafd4 100644
--- a/M2/Macaulay2/d/strings.d
+++ b/M2/Macaulay2/d/strings.d
@@ -36,10 +36,9 @@ export (s:string) >= (t:string) : bool := !(s<t);
 export (s:string) > (t:string) : bool := t<s;
 export (s:string) <= (t:string) : bool := !(t<s);
 export hash(s:string):int := (
-     h := 0;
-     foreach c in s do h = 31*h + int(c);
-     h & 0x7fffffff
-     );
+     h := uint(0);
+     foreach c in s do h = uint(31)*h + uint(c);
+     int(h & 0x7fffffff) );
 
 export index(s:string,offset:int):int := (
      i := offset;
```

**Comment by d-torrance** [comment]
> I think the clearing of the sign bit got optimized away. Get the command line from "make" for compiling that file, add -O0 to it, recompile, relink, and see if the bug is gone.

That worked!

```m2
Macaulay2, version 1.16.0.2
with packages: ConwayPolynomials, Elimination, IntegralClosure, InverseSystems,
               LLLBases, MinimalPrimes, PrimaryDecomposition, ReesAlgebra,
               Saturation, TangentCone, Truncations
--warning: file doc.css not installed yet in package Style
--warning: file katex not installed yet in package Style

i1 : print "Hello from a Raspberry Pi!"
Hello from a Raspberry Pi!
````

**Comment by DanGrayson** [comment]
Did you use the patch or just not optimize?  If the patch works, it's better.

**Comment by d-torrance** [comment]
> Did you use the patch or just not optimize? If the patch works, it's better.

`-O0` first, and then the patch with optimization.  Both worked!  I'll confirm that the rest of the build works and then submit a PR.

**Comment by DanGrayson** [comment]
Great!  I did some tests at this end, and hash codes of strings seem to be unchanged.

**Comment by DanGrayson** [comment]
There are several other places with 0x7fffffff -- could you patch all of them (if they have overflows)?

**Comment by d-torrance** [comment]
> There are several other places with 0x7fffffff -- could you patch all of them (if they have overflows)?

Several of these (in `varstrin.d` and `varnets.d`) are nearly identically to the one in `strings.d` -- hashing functions for string-like classes.  So these seem like no-brainers to switch over to the same behavior.

However, there's a couple that do this to the output of `mpz_get_si` [1] (or `toInt`, which is a wrapper around this).  This returns a signed **long**.  On 32-bit systems, where ints and longs are the same, there's no chance of overflow, but on 64-bit systems, this might be a problem.    But this function deals explicitly with signed ints, so I'm not sure if moving over to unsigned temporarily is the right approach.  Thoughts?

https://github.com/Macaulay2/M2/blob/8ecbb08b154be4e1a39e83faf8576112311e322d/M2/Macaulay2/d/gmp.d#L131-L134
https://github.com/Macaulay2/M2/blob/cbd7cca80b55d071589f77b97a2a3251a4dd2c21/M2/Macaulay2/e/interface/random.cpp#L25-L33

[1] https://gmplib.org/manual/Converting-Integers

**Comment by DanGrayson** [comment]
> I just noticed that all of the architectures that are segfaulting have unsigned `char`, while the architectures where Macaulay2 builds (or at least, the build doesn't fail until the example-generating/testing phase...) have signed `char`. Perhaps that's the issue?

It turns out that's exactly the issue.  The computation of the hash code of a string, if done with unsigned chars, results in a positive value, unless overflow occurs, so by the c11 standard, clearing the sign bit can be omitted by the compiler.

A better fix is for me to modify the d-to-c translator to change "char" to "signed char", for uniformity across compilers.

(The definition of "string" translates to this:
```m2
struct M2_string_struct {int len;char array[1];};
```
).

**Comment by d-torrance** [comment]
> A better fix is for me to modify the d-to-c translator to change "char" to "signed char", for uniformity across compilers.

Sounds good!  I'll hold off on submitting a PR.

(The build of a Debian package on the Raspberry Pi has been chugging along since last night.  Not quite done yet, but no errors so far!)

**Comment by DanGrayson** [comment]
```
commit f67b48278189dcbe2d24b799cfc74b71a0447698 (HEAD -> development)
Author: Daniel R. Grayson <dan@math.uiuc.edu>
Date:   Wed Dec 9 06:51:22 2020 -0600

    fix a problem about whether "char" is signed
    
    Most C compilers make "char" signed but some make it unsigned.  This results in differences
    acrosss compilers, so here we arrange for the datatype "char" in the D language to
    translate to "signed char" in C.  We also had some tests to ensure the results persists.
```

**Comment by d-torrance** [comment]
Closing, thanks to f67b482 and be70ed2.
